// @ngneat/overview/bundles/ngneat-overview.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('rxjs'), require('rxjs/operators')) :
    typeof define === 'function' && define.amd ? define('@ngneat/overview', ['exports', '@angular/core', 'rxjs', 'rxjs/operators'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.ngneat = global.ngneat || {}, global.ngneat.overview = {}), global.ng.core, global.rxjs, global.rxjs.operators));
}(this, (function (exports, i0, rxjs, operators) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () {
                            return e[k];
                        }
                    });
                }
            });
        }
        n['default'] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

    var TeleportService = /** @class */ (function () {
        function TeleportService() {
            this.outlets = new rxjs.BehaviorSubject('');
            this.asObservable = this.outlets.asObservable();
            this.ports = new Map();
        }
        TeleportService.prototype.outlet$ = function (name) {
            var _this = this;
            return this.asObservable.pipe(operators.filter(function (current) { return current === name; }), operators.map(function (name) { return _this.ports.get(name); }));
        };
        TeleportService.prototype.newOutlet = function (name) {
            this.outlets.next(name);
        };
        return TeleportService;
    }());
    TeleportService.ɵprov = i0__namespace.ɵɵdefineInjectable({ factory: function TeleportService_Factory() { return new TeleportService(); }, token: TeleportService, providedIn: "root" });
    TeleportService.decorators = [
        { type: i0.Injectable, args: [{
                    providedIn: 'root',
                },] }
    ];

    var TeleportOutletDirective = /** @class */ (function () {
        function TeleportOutletDirective(vcr, service) {
            this.vcr = vcr;
            this.service = service;
        }
        TeleportOutletDirective.prototype.ngOnInit = function () {
            this.service.ports.set(this.teleportOutlet, this.vcr);
            this.service.newOutlet(this.teleportOutlet);
        };
        TeleportOutletDirective.prototype.ngOnDestroy = function () {
            this.service.ports.delete(this.teleportOutlet);
        };
        return TeleportOutletDirective;
    }());
    TeleportOutletDirective.decorators = [
        { type: i0.Directive, args: [{
                    selector: '[teleportOutlet]',
                },] }
    ];
    TeleportOutletDirective.ctorParameters = function () { return [
        { type: i0.ViewContainerRef },
        { type: TeleportService }
    ]; };
    TeleportOutletDirective.propDecorators = {
        teleportOutlet: [{ type: i0.Input }]
    };

    var TeleportDirective = /** @class */ (function () {
        function TeleportDirective(tpl, service) {
            this.tpl = tpl;
            this.service = service;
        }
        TeleportDirective.prototype.ngOnInit = function () {
            var _this = this;
            this.subscription = this.service.outlet$(this.teleportTo).subscribe(function (outlet) {
                if (outlet) {
                    _this.viewRef = outlet.createEmbeddedView(_this.tpl);
                }
            });
        };
        TeleportDirective.prototype.ngOnDestroy = function () {
            var _a, _b;
            (_a = this.subscription) === null || _a === void 0 ? void 0 : _a.unsubscribe();
            (_b = this.viewRef) === null || _b === void 0 ? void 0 : _b.destroy();
        };
        return TeleportDirective;
    }());
    TeleportDirective.decorators = [
        { type: i0.Directive, args: [{
                    selector: '[teleportTo]',
                },] }
    ];
    TeleportDirective.ctorParameters = function () { return [
        { type: i0.TemplateRef },
        { type: TeleportService }
    ]; };
    TeleportDirective.propDecorators = {
        teleportTo: [{ type: i0.Input }]
    };
    var TeleportModule = /** @class */ (function () {
        function TeleportModule() {
        }
        return TeleportModule;
    }());
    TeleportModule.decorators = [
        { type: i0.NgModule, args: [{
                    declarations: [TeleportDirective, TeleportOutletDirective],
                    exports: [TeleportDirective, TeleportOutletDirective],
                },] }
    ];

    function isTemplateRef(value) {
        return value instanceof i0.TemplateRef;
    }
    function isComponent(value) {
        return typeof value === 'function';
    }
    function isString(value) {
        return typeof value === 'string';
    }

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from) {
        for (var i = 0, il = from.length, j = to.length; i < il; i++, j++)
            to[j] = from[i];
        return to;
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var TplRef = /** @class */ (function () {
        function TplRef(args) {
            this.args = args;
            this.wrapper = null;
            if (this.args.vcr) {
                this.viewRef = this.args.vcr.createEmbeddedView(this.args.tpl, this.args.context || {});
                this.viewRef.detectChanges();
            }
            else {
                this.viewRef = this.args.tpl.createEmbeddedView(this.args.context || {});
                this.viewRef.detectChanges();
                this.args.appRef.attachView(this.viewRef);
            }
        }
        TplRef.prototype.detectChanges = function () {
            this.viewRef.detectChanges();
        };
        TplRef.prototype.getElement = function () {
            var _b;
            var rootNodes = this.viewRef.rootNodes;
            if (rootNodes.length === 1 && rootNodes[0] === Node.ELEMENT_NODE) {
                this.element = rootNodes[0];
            }
            else {
                this.element = document.createElement('div');
                (_b = this.element).append.apply(_b, __spreadArray([], __read(rootNodes)));
            }
            return this.element;
        };
        TplRef.prototype.destroy = function () {
            var _a;
            if (this.viewRef.rootNodes[0] !== 1) {
                (_a = this.element) === null || _a === void 0 ? void 0 : _a.parentNode.removeChild(this.element);
                this.element = null;
            }
            if (!this.args.vcr) {
                this.args.appRef.detachView(this.viewRef);
            }
            this.viewRef.destroy();
            this.viewRef = null;
        };
        return TplRef;
    }());

    var StringRef = /** @class */ (function () {
        function StringRef(value) {
            this.value = value;
        }
        StringRef.prototype.getElement = function () {
            return this.value;
        };
        StringRef.prototype.detectChanges = function () { };
        StringRef.prototype.destroy = function () { };
        return StringRef;
    }());

    var CompRef = /** @class */ (function () {
        function CompRef(args) {
            this.args = args;
            var factory = this.args.resolver.resolveComponentFactory(this.args.component);
            if (this.args.vcr) {
                this.compRef = this.args.vcr.createComponent(factory, this.args.vcr.length, args.injector || this.args.vcr.injector);
            }
            else {
                this.compRef = factory.create(this.args.injector);
                this.args.appRef.attachView(this.compRef.hostView);
            }
        }
        Object.defineProperty(CompRef.prototype, "ref", {
            get: function () {
                return this.compRef;
            },
            enumerable: false,
            configurable: true
        });
        CompRef.prototype.setInput = function (input, value) {
            this.compRef.instance[input] = value;
            return this;
        };
        CompRef.prototype.setInputs = function (inputs) {
            var _this = this;
            Object.keys(inputs).forEach(function (input) {
                _this.compRef.instance[input] = inputs[input];
            });
            return this;
        };
        CompRef.prototype.detectChanges = function () {
            this.compRef.hostView.detectChanges();
            return this;
        };
        CompRef.prototype.appendTo = function (container) {
            container.appendChild(this.getElement());
            return this;
        };
        CompRef.prototype.removeFrom = function (container) {
            container.removeChild(this.getElement());
            return this;
        };
        CompRef.prototype.getRawContent = function () {
            return this.getElement().outerHTML;
        };
        CompRef.prototype.getElement = function () {
            return this.compRef.location.nativeElement;
        };
        CompRef.prototype.destroy = function () {
            this.compRef.destroy();
            !this.args.vcr && this.args.appRef.detachView(this.compRef.hostView);
            this.compRef = null;
        };
        return CompRef;
    }());

    var ViewService = /** @class */ (function () {
        function ViewService(resolver, injector, appRef) {
            this.resolver = resolver;
            this.injector = injector;
            this.appRef = appRef;
        }
        ViewService.prototype.createComponent = function (component, options) {
            if (options === void 0) { options = {}; }
            return new CompRef({
                component: component,
                vcr: options.vcr,
                injector: options.injector || this.injector,
                appRef: this.appRef,
                resolver: this.resolver,
            });
        };
        ViewService.prototype.createTemplate = function (tpl, options) {
            if (options === void 0) { options = {}; }
            return new TplRef({
                vcr: options.vcr,
                appRef: this.appRef,
                tpl: tpl,
                context: options.context,
            });
        };
        ViewService.prototype.createView = function (content, viewOptions) {
            if (viewOptions === void 0) { viewOptions = {}; }
            if (isTemplateRef(content)) {
                return this.createTemplate(content, viewOptions);
            }
            else if (isComponent(content)) {
                return this.createComponent(content, viewOptions);
            }
            else if (isString(content)) {
                return new StringRef(content);
            }
            else {
                throw 'Type of content is not supported';
            }
        };
        return ViewService;
    }());
    ViewService.ɵprov = i0__namespace.ɵɵdefineInjectable({ factory: function ViewService_Factory() { return new ViewService(i0__namespace.ɵɵinject(i0__namespace.ComponentFactoryResolver), i0__namespace.ɵɵinject(i0__namespace.INJECTOR), i0__namespace.ɵɵinject(i0__namespace.ApplicationRef)); }, token: ViewService, providedIn: "root" });
    ViewService.decorators = [
        { type: i0.Injectable, args: [{ providedIn: 'root' },] }
    ];
    ViewService.ctorParameters = function () { return [
        { type: i0.ComponentFactoryResolver },
        { type: i0.Injector },
        { type: i0.ApplicationRef }
    ]; };

    var DynamicViewComponent = /** @class */ (function () {
        function DynamicViewComponent() {
        }
        return DynamicViewComponent;
    }());
    DynamicViewComponent.decorators = [
        { type: i0.Component, args: [{
                    selector: 'dynamic-view',
                    template: " <div [innerHTML]=\"content\"></div> "
                },] }
    ];
    DynamicViewComponent.propDecorators = {
        content: [{ type: i0.Input }]
    };

    var DynamicViewDirective = /** @class */ (function () {
        function DynamicViewDirective(defaultTpl, vcr, viewService) {
            this.defaultTpl = defaultTpl;
            this.vcr = vcr;
            this.viewService = viewService;
        }
        Object.defineProperty(DynamicViewDirective.prototype, "view", {
            get: function () {
                return this._view;
            },
            set: function (content) {
                this._view = content;
                this.resolveContentType();
            },
            enumerable: false,
            configurable: true
        });
        DynamicViewDirective.prototype.ngOnInit = function () {
            this.resolveContentType();
        };
        DynamicViewDirective.prototype.resolveContentType = function () {
            var _a;
            (_a = this.viewRef) === null || _a === void 0 ? void 0 : _a.destroy();
            if (isString(this.view)) {
                this.viewRef = this.viewService.createComponent(DynamicViewComponent, {
                    vcr: this.vcr,
                    injector: this.injector,
                });
                this.viewRef.setInput('content', this.view).detectChanges();
            }
            else {
                this.viewRef = this.viewService.createView(this.view || this.defaultTpl, {
                    vcr: this.vcr,
                    injector: this.injector,
                    context: this.context,
                });
            }
        };
        DynamicViewDirective.prototype.ngOnDestroy = function () {
            var _a;
            (_a = this.viewRef) === null || _a === void 0 ? void 0 : _a.destroy();
        };
        return DynamicViewDirective;
    }());
    DynamicViewDirective.decorators = [
        { type: i0.Directive, args: [{
                    selector: '[dynamicView]',
                },] }
    ];
    DynamicViewDirective.ctorParameters = function () { return [
        { type: i0.TemplateRef },
        { type: i0.ViewContainerRef },
        { type: ViewService }
    ]; };
    DynamicViewDirective.propDecorators = {
        view: [{ type: i0.Input, args: ['dynamicView',] }],
        injector: [{ type: i0.Input, args: ['dynamicViewInjector',] }],
        context: [{ type: i0.Input, args: ['dynamicViewContext',] }]
    };
    var DynamicViewModule = /** @class */ (function () {
        function DynamicViewModule() {
        }
        return DynamicViewModule;
    }());
    DynamicViewModule.decorators = [
        { type: i0.NgModule, args: [{
                    declarations: [DynamicViewDirective, DynamicViewComponent],
                    exports: [DynamicViewDirective],
                },] }
    ];

    /*
     * Public API Surface of overview
     */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.CompRef = CompRef;
    exports.DynamicViewDirective = DynamicViewDirective;
    exports.DynamicViewModule = DynamicViewModule;
    exports.StringRef = StringRef;
    exports.TeleportDirective = TeleportDirective;
    exports.TeleportModule = TeleportModule;
    exports.TeleportOutletDirective = TeleportOutletDirective;
    exports.TplRef = TplRef;
    exports.ViewService = ViewService;
    exports.isComponent = isComponent;
    exports.isString = isString;
    exports.isTemplateRef = isTemplateRef;
    exports.ɵa = TeleportService;
    exports.ɵb = DynamicViewComponent;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
